src/originalBudgetTrackingApp/eventEmitters/EventEmitter.ts

```diff
@@ -0,0 +1,76 @@
+import Emittery from 'emittery';
+import { AccountToScrapeConfig } from '@/originalBudgetTrackingApp/configManager/configManager';
+import { EnrichedTransaction } from '@/originalBudgetTrackingApp/commonTypes';
+
+export enum EventNames {
+  IMPORT_PROCESS_START = 'IMPORT_PROCESS_START',
+  IMPORTER_START = 'IMPORTER_START',
+  IMPORTER_PROGRESS = 'IMPORTER_PROGRESS',
+  IMPORTER_ERROR = 'IMPORTER_ERROR',
+  IMPORTER_END = 'IMPORTER_END',
+  IMPORT_PROCESS_END = 'IMPORT_PROCESS_END',
+  EXPORT_PROCESS_START = 'EXPORT_PROCESS_START',
+  EXPORTER_START = 'EXPORTER_START',
+  EXPORTER_PROGRESS = 'EXPORTER_PROGRESS',
+  EXPORTER_ERROR = 'EXPORTER_ERROR',
+  EXPORTER_END = 'EXPORTER_END',
+  EXPORT_PROCESS_END = 'EXPORT_PROCESS_END',
+  GENERAL_ERROR = 'GENERAL_ERROR',
+  LOG = 'LOG'
+}
+
+interface ErrorEvent {
+  error: Error
+}
+
+interface BudgetTrackingEvent {
+  message?: string;
+}
+
+interface ImporterEvent extends BudgetTrackingEvent {
+  id: string
+  name: string
+  companyKey: AccountToScrapeConfig['key']
+}
+
+interface ImporterErrorEvent extends ImporterEvent, ErrorEvent {
+}
+
+interface ImporterEndEvent extends ImporterEvent {
+ transactions: EnrichedTransaction[]
+}
+
+interface ExporterEvent extends BudgetTrackingEvent {
+  name: string
+  allTransactions: EnrichedTransaction[]
+}
+
+interface ExporterErrorEvent extends ExporterEvent, ErrorEvent {
+
+}
+
+interface ImportProcessStartEvent extends BudgetTrackingEvent {
+  startDate: Date
+}
+
+type EventDataMap = {
+  [EventNames.IMPORT_PROCESS_START]: ImportProcessStartEvent
+  [EventNames.IMPORTER_START]: ImporterEvent
+  [EventNames.IMPORTER_PROGRESS]: ImporterEvent
+  [EventNames.IMPORTER_ERROR]: ImporterErrorEvent
+  [EventNames.IMPORTER_END]: ImporterEndEvent
+  [EventNames.IMPORT_PROCESS_END]: { }
+  [EventNames.EXPORT_PROCESS_START]: { }
+  [EventNames.EXPORTER_START]: ExporterEvent
+  [EventNames.EXPORTER_PROGRESS]: ExporterEvent
+  [EventNames.EXPORTER_ERROR]: ExporterErrorEvent
+  [EventNames.EXPORTER_END]: ExporterEvent
+  [EventNames.GENERAL_ERROR]: ErrorEvent
+  [EventNames.LOG]: BudgetTrackingEvent
+};
```

Comment:
I just found a better way to create an `enum` like this.
You can create your type, for example

```ts

type EventDataMap = {
  IMPORT_PROCESS_START: ImportProcessStartEvent
  IMPORTER_START: ImporterEvent
  ...
};
```

Then to export the keys as enum you can
```ts
export type EventNames = keyof EventDataMap
```

Reply:
That was my initial thought as well. But the result of `keyof EventDataMap` is a union of all the strings of the keys. It does not behave like an enum and doesn't allow me to use the `EventNames.xxxxxxx` to get a specific value.
![Screen Shot 2020-09-14 at 15 26 24](https://user-images.githubusercontent.com/7272927/93086104-45d45e00-f69f-11ea-8c4d-79564bd2a41d.png)

Reply:
😌I remember `keyof` return an `enum`, my mistake